server/snapshot/snapshot.go

```diff
@@ -0,0 +1,161 @@
+func GetOrCreateSnapshotKey(gun string, store storage.KeyStore, crypto signed.CryptoService, createAlgorithm string) (data.PublicKey, error) {
+	keyAlgorithm, public, err := store.GetKey(gun, data.CanonicalSnapshotRole)
+	if err == nil {
+		return data.NewPublicKey(keyAlgorithm, public), nil
+	}
+
+	if _, ok := err.(*storage.ErrNoKey); ok {
+		key, err := crypto.Create("snapshot", createAlgorithm)
+		if err != nil {
+			return nil, err
+		}
+		logrus.Debug("Creating new snapshot key for ", gun, ". With algo: ", key.Algorithm())
+		err = store.SetKey(gun, data.CanonicalSnapshotRole, key.Algorithm(), key.Public())
+		if err == nil {
+			return key, nil
+		}
+
+		if _, ok := err.(*storage.ErrKeyExists); ok {
+			keyAlgorithm, public, err = store.GetKey(gun, data.CanonicalSnapshotRole)
+			if err != nil {
+				return nil, err
+			}
+			return data.NewPublicKey(keyAlgorithm, public), nil
+		}
+		return nil, err
+	}
+	return nil, err
+}
+
+// GetOrCreateSnapshot either returns the exisiting latest snapshot, or uses
+// whatever the most recent snapshot is to generate a new one.
+func GetOrCreateSnapshot(gun string, store storage.MetaStore, cryptoService signed.CryptoService) ([]byte, error) {
+
+	d, err := store.GetCurrent(gun, "snapshot")
+	if err != nil {
+		return nil, err
+	}
+
+	sn := &data.SignedSnapshot{}
+	if d != nil {
+		err := json.Unmarshal(d, sn)
+		if err != nil {
+			logrus.Error("Failed to unmarshal existing snapshot")
+			return nil, err
+		}
+		if !snapshotExpired(sn) && !contentExpired(gun, sn, store) {
+			return d, nil
+		}
+	}
+
+	sgnd, version, err := createSnapshot(gun, sn, store, cryptoService)
+	if err != nil {
+		logrus.Error("Failed to create a new snapshot")
+		return nil, err
+	}
+	out, err := json.Marshal(sgnd)
+	if err != nil {
+		logrus.Error("Failed to marshal new snapshot")
+		return nil, err
+	}
+	err = store.UpdateCurrent(gun, storage.MetaUpdate{Role: "snapshot", Version: version, Data: out})
+	if err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+// snapshotExpired simply checks if the snapshot is past its expiry time
+func snapshotExpired(sn *data.SignedSnapshot) bool {
+	return signed.IsExpired(sn.Signed.Expires)
+}
+
+// contentExpired checks to see if any of the roles already in the snapshot
+// have been updated. It will update any roles that have changed as it goes
+// so that we don't have to run through all this again a second time.
+func contentExpired(gun string, sn *data.SignedSnapshot, store storage.MetaStore) bool {
+	expired := false
+	updatedMeta := make(data.Files)
+	for role, meta := range sn.Signed.Meta {
+		curr, err := store.GetCurrent(gun, role)
+		if err != nil {
+			return false
+		}
+		roleExp, newHash := roleExpired(curr, meta)
+		if roleExp {
+			updatedMeta[role] = data.FileMeta{
+				Length: int64(len(curr)),
+				Hashes: data.Hashes{
+					"sha256": newHash,
+				},
+			}
+		}
+		expired = expired || roleExp
+	}
+	if expired {
+		sn.Signed.Meta = updatedMeta
```

Comment:
(I appreciate that computing the hash only once is clever, but this seems like a place where not being clever and calling some kind of canonical/non-repeated implementation in `tuf.Repo` might be worthwhile.)

And more generally, how can this function ever return `true`? AFAICS if the snapshot key is on the server (i.e. not available to the clients), the clients will not be able to send a snapshot during any update, so the server will auto-generate the snapshot at that time, and update all of them within a transaction. So AFAICS this could only trigger with a client which defines two snapshot keys, one kept locally and one on the server. What scenario am I missing?